src/chart/custom/CustomView.ts

```diff
@@ -1316,24 +1322,43 @@ function mergeChildren(
     // might be better performance.
     let index = 0;
     for (; index < newLen; index++) {
-        newChildren[index] && doCreateOrUpdateEl(
-            api,
-            el.childAt(index),
-            dataIndex,
-            newChildren[index] as CustomElementOption,
-            seriesModel,
-            el
-        );
+        const newChild = newChildren[index];
+        const oldChild = el.childAt(index);
+        if (newChild) {
+            doCreateOrUpdateEl(
+                api,
+                oldChild,
+                dataIndex,
+                newChild as CustomElementOption,
+                seriesModel,
+                el
+            );
+        }
+        else {
+            removeChildFromGroup(el, oldChild, seriesModel);
```

Comment:
If leave animation is disabled, it will still call `parent.remove(el)` directly, which uses `splice` internally and brings a weird behavior:

existing children: `[elA, elB, elC]`
new children: `[elM, null, elN]`
result: `[merge(elA, elM), elC, elN]`


I've changed my opinion that
> Basically I agree this change, which might make it more intuitive

When setOption, the default strategy is "merge", which is the same as the other series like line, scatter, ...
The merge strategy in custom series contains: 
+ dataItems are merged by index
+ group children are merge by index

In series link line, scatter, graphics are the same between different data items, and it makes "merge" strategy make sense in most cases. But is custom series, graphics may not be the same between different data items, which probably make the "merge" not reasonable. 

This change brings a breaking change but only change the `null` case in group children but not considered the case like "merge graphic unexpectedly of different data item", "remain properties unexpectedly when merge graphic". I think it is not good enough.

Could we provide a option on custom series like 
```js
series: {
     type: 'custom',
     merge: false,
}
```
When `merge: false`, there is no merge either between different data items (see https://github.com/apache/echarts/blob/5.3.3/src/chart/custom/CustomView.ts#L222)
or in group children.

And it may fix #17349 thoroughly ?

@Ovilia @pissang what's your opinion?

Reply:
I think this mainly depends on how we understand *merging*. I don't think it's intuitive to use the old element in a group when provided `null` in new `setOption`. This is just like when providing series data like `[10, 20, null, 40]` in a new `setOption` should not take the null data to use the old value in the previous data.

Reply:
You are quite right! I didn't think of the situation. I fixed this and also update the test case.

Reply:
After adding `oldRemovedCount` the effect is better but still has some issues:

## Consider the case below:
+ step 1. `setOption` with new children: `[elA, elB, elC]` (length: 3)
+ step 2. `setOption` with new children: `[elA, null, elC]` (length: 3) (means hide `elB`)
+ step 3. `setOption` with new children: `[elA, elB, elC]` (length: 3) (means show `elB` again)

But with current modified code, the result will be:
+ After step 1. children will be `[elA, elB, elC]`
+ After step 2. children will be `[elA, elC]`
+ After step 3. children will be `[elA, merge(elC, elB), elC]`

which is unexpected.

## Consider leave animation
If the `setOption` is triggered by user's repeat click, the interval between two clicks is not predicable, may be longer or shorter than the leave animation duration. In this case, if the indices of some children is different before and after leave animation, the merge result will be not predicable.

---

Basically, if we use `null/undefined` to delete a child, I think this principle should be apply: 
The index of any other elements should not be changed after this deletion.
That is, is we delete a child, that place may be kept "no element" or "an placeholder element that can not displayed"